sympy/core/tests/test_equation.py

```diff
@@ -0,0 +1,122 @@
+    diff, FiniteSet, Matrix, exp, collect
+from sympy import Equality
+from sympy import Equation, Eqn
+from sympy import latex
+from sympy.testing.pytest import raises, warns
+
+
+def test_define_equation():
+    a, b, c = var('a b c')
+    raises(TypeError, lambda: Eqn(FiniteSet(a), FiniteSet(b, c)))
+    with warns(UserWarning):
+        assert Eqn(1, 0) == Equation(1, 0, check=False)
+    # No warning with `check=False`.
+    assert Eqn(1, 0, check=False) == Equation(1, 0, check=False)
+    tsteqn = Eqn(a, b / c)
+    assert tsteqn == Equation(a, b / c)
+    assert tsteqn.lhs == Equation(a, b / c).args[0]
+    assert tsteqn.rhs == Equation(a, b / c).args[1]
+    assert tsteqn.free_symbols == {a, b, c}
+
+
+def test_convert_equation():
+    a, b, c = var('a b c')
+    tsteqn = Eqn(a, b / c)
+    assert tsteqn.as_Boolean() == Equality(a, b / c)
+    assert tsteqn.reversed == Equation(b / c, a)
+    assert tsteqn.swap == Equation(b / c, a)
+
+
+def test_binary_op():
+    a, b, c = var('a b c')
+    tsteqn = Eqn(a, b / c)
+    assert tsteqn + c == Equation(a + c, b / c + c)
+    assert c + tsteqn == Equation(c + a, c + b / c)
+    assert tsteqn * c == Equation(a * c, b)
+    assert c * tsteqn == Equation(c * a, b)
+    assert tsteqn - c == Equation(a - c, b / c - c)
+    assert c - tsteqn == Equation(c - a, c - b / c)
+    assert tsteqn / c == Equation(a / c, b / c ** 2)
+    assert c / tsteqn == Equation(c / a, c ** 2 / b)
+    assert tsteqn % c == Equation(a % c, (b / c) % c)
+    assert c % tsteqn == Equation(c % a, c % (b / c))
+    assert tsteqn ** c == Equation(a ** c, (b / c) ** c)
+    assert c ** tsteqn == Equation(c ** a, c ** (b / c))
+    assert tsteqn + tsteqn == Equation(2 * a, 2 * b / c)
+    assert tsteqn * tsteqn == Equation(a ** 2, b ** 2 / c ** 2)
+    assert tsteqn - tsteqn == Equation(0, 0)
+    assert tsteqn / tsteqn == Equation(1, 1)
+    assert tsteqn % tsteqn == Equation(0, 0)
+    assert tsteqn ** tsteqn == Equation(a ** a, (b / c) ** (b / c))
+
+
+def test_outputs():
+    a, b, c = var('a b c')
+    tsteqn = Eqn(a, b / c)
+    assert str(tsteqn) == 'a = b/c'
+    assert latex(tsteqn) == 'a = \\frac{b}{c}'
+
+
+def test_helper_functions():
+    a, b, c = var('a b c')
+    tsteqn = Eqn(a, b / c)
+    raises(ValueError, lambda: integrate(tsteqn, c))
+    raises(AttributeError, lambda: integrate(tsteqn, c, side='right'))
+    assert tsteqn.evalf(4, {b: 2.0, c: 4}) == Eqn(a, 0.5000)
+    assert diff(tsteqn, c) == Equation(diff(a, c, evaluate=False), -b / c ** 2)
+    tsteqn = Eqn(a * c, b / c)
+    assert diff(tsteqn, c) == Equation(a, -b / c ** 2)
+    assert integrate(tsteqn, c, side='rhs') == integrate(tsteqn.rhs, c)
+    assert integrate(tsteqn, c, side='lhs') == integrate(tsteqn.lhs, c)
+
+    def adsq(expr):
+        # Arbitrary python function
+        return expr + expr ** 2
+
+    assert (adsq(tsteqn) == Eqn(a ** 2 * c ** 2 + a * c, b ** 2 / c ** 2 +
+                                b / c))
```

Comment:
I find this hard to read. I would prefer:
```
    assert (adsq(tsteqn) == Eqn(a**2*c**2 + a*c, b**2/c**2 + b/c))
```

Reply:
Actually the outer brackets are not needed:
```
    assert adsq(tsteqn) == Eqn(a**2*c**2 + a*c, b**2/c**2 + b/c)
```

Reply:
The spaces are part of PEP8. I will try removing them and see if it passes format checking. I completely agree with you on readability.

Reply:
We don't need to follow PEP8 but in any case it says:
```
If operators with different priorities are used, consider adding whitespace around the
operators with the lowest priority(ies). Use your own judgment; however, never use more
than one space, and always have the same amount of whitespace on both sides of a
binary operator:

# Correct:
i = i + 1
submitted += 1
x = x*2 - 1
hypot2 = x*x + y*y
c = (a+b) * (a-b)
```

Reply:
@oscarbenjamin thanks for tracking down the exact info for me.

Reply:
fixed.
commit 4dc62ee7